File: Assignment_5_Baseline/Users/Anthony_User.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import os
import time

# Set the base URLs for the websites
url_with_keyword = 'https://en.wikipedia.org/wiki/Web_scraping'
url_without_keyword = 'https://www.example.com/without-images'
url_test = 'http://localhost:3000/'

# Set the keyword to search for
keyword = 'web scraping'

# Path to save screenshots
images_directory = './Images'
os.makedirs(images_directory, exist_ok=True)

def run_test(url, has_keyword):
    driver = None
    try:
        # Initialize the Chrome WebDriver
        driver = webdriver.Chrome()

        # Open the URL
        driver.get(url)

        # Wait for the page to fully load
        WebDriverWait(driver, 30).until(EC.presence_of_element_located((By.TAG_NAME, 'body')))

        presence_time = 0
        found_keyword = False
        found_image = False
        found_link = False

        # Continuously check if the content contains the keyword
        while presence_time < 20:
            content = driver.find_element(By.TAG_NAME, 'body').text
            if keyword in content:
                found_keyword = True
                presence_time += 10
                break

            time.sleep(1)
            presence_time += 1

        # Check for images on the page
        images = driver.find_elements(By.TAG_NAME, 'img')
        if images:
            found_image = True
            for image in images:
                time.sleep(10)
                presence_time += 10

        # Check for links on the page
        link = driver.find_element(By.TAG_NAME, 'a')
        if link:
            found_link = True
            # Additional checks and navigation can be included here

            # Take action on the link
            link.click()
            WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.TAG_NAME, 'body')))

        # Print results
        print(f'Presence time on {url}: {presence_time} seconds')
        print(f'Keyword found: {found_keyword}')
        print(f'Image found: {found_image}')
        print(f'Link found: {found_link}')

        # Save screenshots based on condition
        screenshot_filename = os.path.join(images_directory, 'with-keyword.png' if has_keyword else 'without-keyword.png')
        driver.get_screenshot_as_file(screenshot_filename)

    except Exception as e:
        print(e)
    finally:
        if driver:
            driver.quit()

# Run the tests
run_test(url_test, False)
# run_test(url_with_keyword, True)
# run_test(url_without_keyword, False)

File: Assignment_5_Baseline/Users/Anthony_User.js
const { Builder, By, Key, until } = require('selenium-webdriver');
const fs = require('fs');
const path = require('path');

// Set the base URL for the website with the keyword
const urlWithKeyword = 'https://en.wikipedia.org/wiki/Web_scraping';

// Set the base URL for the website without the keyword
const urlWithoutKeyword = 'https://www.example.com/without-images';

const urlTest = 'http://localhost:3000/';

async function runTest(url, hasKeyword) {
  let driver;
  try {
    driver = await new Builder().forBrowser('chrome').build();

    await driver.get(url);

    // Wait for the page to fully load before checking for elements
    await driver.wait(until.elementLocated(By.tagName('body')), 30000);

    let presenceTime = 0;
    let foundKeyword = false;
    let foundImage = false;
    let foundLink = false;

    // Check if the content contains the keyword every second
    const checkInterval = setInterval(async () => {
      const content = await driver.findElement(By.tagName('body')).getAttribute('innerText');
      if (content.includes(keyword)) {
        clearInterval(checkInterval);
        foundKeyword = true;

        // Extend presence time by 10 seconds if the keyword is found
        presenceTime += 10;
      }

      presenceTime++;

      if (presenceTime >= 20) {
        clearInterval(checkInterval);
      }
    }, 1000);

    // Check if there are any images on the page
    const images = await driver.findElements(By.tagName('img'));
    if (images.length > 0) {
      foundImage = true;

      // Extend presence time by 10 seconds for each image
      for (const image of images) {
        await driver.sleep(10000);
        presenceTime += 10;
      }
    }

    // Check if there is a link on the page
    const link = await driver.findElement(By.tagName('a'));
    if (link) {
      foundLink = true;

      // Wait for the intercepting element to disappear before clicking on the link
      await driver.wait(until.elementLocated(By.css('.vector-menu-toggle')), 10000).then(() => {
        driver.findElement(By.css('.vector-menu-toggle')).click();
      });

      // Extend presence time by 10 seconds for the link
      await driver.sleep(10000);
      presenceTime += 10;

      // Click on the link and repeat the process
      await link.click();
      await driver.wait(until.elementLocated(By.tagName('body')), 10000);
      await runTest(driver.getCurrentUrl(), hasKeyword);
    }

    console.log(`Presence time on ${url}: ${presenceTime} seconds`);
    console.log(`Keyword found: ${foundKeyword}`);
    console.log(`Image found: ${foundImage}`);
    console.log(`Link found: ${foundLink}`);

    if (hasKeyword) {
      await driver.takeScreenshot().then(buffer => {
        fs.writeFileSync(path.join(__dirname, 'Images', 'with-keyword.png'), buffer);
      });
    } else {
      await driver.takeScreenshot().then(buffer => {
        fs.writeFileSync(path.join(__dirname, 'Images', 'without-keyword.png'), buffer);
      });
    }
  } catch (err) {
    console.error(err);
  } finally {
    if (driver) {
      await driver.quit();
    }
  }
}

// Set the keyword to search for
const keyword = 'web scraping';
try {

runTest(urlTest, false)

// Run the test against the website with the keyword
runTest(urlWithKeyword, true);

// Run the test against the website without the keyword
runTest(urlWithoutKeyword, false);
} catch (err) {
    console.error(err);

}
